cmd/maya-exporter/app/collector/cstor.go

```diff
@@ -0,0 +1,211 @@
+
+// reader reads the response from socket in the buffer and
+// exits if the buffer contains the "IOSTATS\r\n" at the end.
+func (c *cstor) reader() (string, error) {
+	buf := make([]byte, BufSize)
+	var (
+		err           error
+		n             int
+		str, response string
+		buffer        bytes.Buffer
+	)
+	// infinite for loop to collect all the chunks.
+	for {
+		n, err = c.conn.Read(buf[:])
+		if err != nil {
+			return "", err
+		}
+		// concatnat the chunks received and then compare if it has
+		// Footer (" IOSTATS") + EOF ("\r\n") and exit, else continue
+		// appending the chunks at the end of str.
+		buffer.WriteString(string(buf[0:n]))
+		str = buffer.String()
+		if str[len(str)-12:] == Footer+EOF {
+			response = str
+			break
+		}
+	}
+	return response, nil
+}
+
+// ReadHeader only reads the header of the response from cstor
+func (c *cstor) readHeader() error {
+	buf := make([]byte, 1024)
+	var (
+		err    error
+		n      int
+		str    string
+		buffer bytes.Buffer
+	)
+	// collect all the chunks ending with EOF ("\r\n").
+	for {
+		n, err = c.conn.Read(buf[:])
+		if err != nil {
+			glog.Error("Error in reading response, error : ", err)
+			return err
+		}
+		// apend the chunks into str
+		buffer.WriteString(string(buf[0:n]))
+		str = buffer.String()
+		if strings.HasPrefix(str, HeaderPrefix) && strings.HasSuffix(str, EOF) {
+			break
+		}
+	}
+	glog.Infof("Got header: %#v", str)
+	return nil
+}
+
+// getter writes query on the socket for the getting stats
+// if the connection is available else retry to initiate
+// connection again.
+func (c *cstor) getter() (v1.VolumeStats, error) {
+	var (
+		err   error
+		stats v1.VolumeStats
+	)
+	if c.conn == nil {
+		// initiate the connection again if connection with the istgt closed
+		// due to timeout or some other errors from istgt side.
+		if err := c.InitiateConnection(SocketPath); err != nil {
+			return v1.VolumeStats{}, &connErr{
+				errors.Errorf("%s: %v", "can't initiate connection with istgt", err),
+			}
+		}
+	}
+	if err := c.writer(); err != nil {
+		c.conn.Close()
+		c.conn = nil
+		return v1.VolumeStats{}, &connErr{
+			errors.Errorf("%v: %v", err, "closing connection"),
+		}
+	}
+	response, err := c.reader()
+	if err != nil {
+		c.conn.Close()
+		c.conn = nil
+		return v1.VolumeStats{}, &connErr{
+			errors.Errorf("%v: %v", err, "closing connection"),
+		}
+	}
+	glog.Infof("Got response: %v", response)
+	response = splitter(response)
+	if len(response) == 0 {
+		return v1.VolumeStats{}, errors.New("Got empty response from cstor")
+	}
+
+	// unmarshal the json response into metrics instances.
+	if err, stats = newResponse(response); err != nil {
+		return v1.VolumeStats{}, err
+	}
+
+	stats.Got = true
+	return stats, nil
+}
+
+// Parser can used to parse the json strings into the respective types.
+// TODO: Instead of using two parser methods make it
+// a generic parser that can be used for both jiva and cstor.
+func (c *cstor) parser(volStats v1.VolumeStats) stats {
```

Comment:
How about `parse` as the method name?

Reply:
Perhaps `func(c *cstor) asStats(s v1.VolumeStats) stats {}` is better

Reply:
i think parse would be good as here we are parsing stats from one structure to other.